File: examples/malformed-json-response/python/mock_server.py
"""OpenAI-compatible mock returning valid inner JSON in the content field."""

import http.server
import json
import threading


def _content_payload() -> str:
    return json.dumps({"tool": "lookup_user", "args": {"id": 42}})


class _Handler(http.server.BaseHTTPRequestHandler):
    def do_POST(self):
        if self.path != "/v1/chat/completions":
            self.send_error(404)
            return
        length = int(self.headers.get("Content-Length", 0))
        self.rfile.read(length)
        body = json.dumps(
            {
                "id": "chatcmpl-test",
                "object": "chat.completion",
                "model": "gpt-4o-mini",
                "choices": [
                    {
                        "index": 0,
                        "message": {
                            "role": "assistant",
                            "content": _content_payload(),
                        },
                        "finish_reason": "stop",
                    }
                ],
            }
        ).encode()
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def log_message(self, *_):
        pass


def serve_in_background():
    server = http.server.HTTPServer(("127.0.0.1", 0), _Handler)
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    return server.server_address[1], server.shutdown

File: examples/malformed-json-response/nodejs/mock-server.js
// OpenAI-compatible mock returning valid inner JSON in the content
// field. Under faultkit's malformed-json-response fault, the
// content gets replaced with intentionally invalid JSON.
import http from "node:http";

const innerPayload = JSON.stringify({ tool: "lookup_user", args: { id: 42 } });

export function serveInBackground() {
  const server = http.createServer((req, res) => {
    if (req.url !== "/v1/chat/completions") {
      res.writeHead(404).end();
      return;
    }
    let body = "";
    req.on("data", (c) => (body += c));
    req.on("end", () => {
      const payload = JSON.stringify({
        id: "chatcmpl-test",
        object: "chat.completion",
        model: "gpt-4o-mini",
        choices: [
          {
            index: 0,
            message: { role: "assistant", content: innerPayload },
            finish_reason: "stop",
          },
        ],
      });
      res.writeHead(200, {
        "Content-Type": "application/json",
        "Content-Length": Buffer.byteLength(payload),
      });
      res.end(payload);
    });
  });
  return new Promise((resolve) => {
    server.listen(0, "127.0.0.1", () => {
      resolve({
        port: server.address().port,
        shutdown: () => new Promise((r) => server.close(r)),
      });
    });
  });
}
